tracing-attributes/src/lib.rs

```diff
@@ -212,6 +211,19 @@ pub fn instrument(args: TokenStream, item: TokenStream) -> TokenStream {
     .into()
 }
 
+fn param_names(pat: Pat) -> Box<dyn Iterator<Item=Ident>> {
+    match pat {
+        Pat::Ident(PatIdent { ident, .. }) => Box::new(iter::once(ident)),
+        Pat::Reference(PatReference { pat, .. }) => param_names(*pat),
+        Pat::Struct(PatStruct { fields, .. }) =>
+            Box::new(fields.into_iter().flat_map(|FieldPat { pat, .. }| param_names(*pat))),
+        Pat::Tuple(PatTuple { elems, .. }) => Box::new(elems.into_iter().flat_map(param_names)),
+        Pat::TupleStruct(PatTupleStruct { pat: PatTuple { elems, .. }, .. }) =>
+            Box::new(elems.into_iter().flat_map(param_names)),
+        _ => Box::new(iter::empty()),
```

Comment:
If this case matches, it's probably a syntax error, right? Can any pattern types we don't currently match be used in function params? I think it might be good to have a comment hereif that's true.

Reply:
Hmm. I think that if the remaining cases are something that will cause `rustc` to error anyway, it's probably best for the proc macro _not_ to panic, so that rustc can print the correct diagnostics for the syntax error (rather than a more surprising "proc macro panicked!" message)? But, it might be good to have a comment noting that. Not a blocker though.

Reply:
For some reason I was operating under the assumption that this would be a compiler error *before* the proc macro was invoked.  There is a class of errors that will bail before trying to invoke the proc macro, but clearly it doesn't include this (I double checked just to be sure).  I *think* the exact details would be found in [RFC #550](https://github.com/rust-lang/rfcs/blob/0a46c3372ea23b9e2f5cab52c098de3b2803a053/text/0550-macro-future-proofing.md), but I really don't have the head space to digest that RFC right now (and it's also not at all useful for the purposes of this PR :laughing:).  A comment does seem useful, if for no other reason than to keep others from going down this rabbit hole.

Reply:
I believe the reason this would error after the proc-macro is evaluated is because the macro could rewrite the function definition. Thanks for adding the comment!